File: view/app/src/main/java/com/google/samples/quickstart/searchinapps/view/MainActivity.java
/* Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.samples.quickstart.searchinapps.view;

import static com.google.android.libraries.searchinapps.SearchSuggestionsViewOptions.Layout.CAROUSEL;
import static com.google.android.libraries.searchinapps.SearchSuggestionsViewOptions.Layout.COMPACT_CAROUSEL;
import static com.google.android.libraries.searchinapps.SearchSuggestionsViewOptions.Layout.TILING;

import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RadioGroup;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.lifecycle.ViewModelProvider;
import com.google.android.libraries.searchinapps.GetSearchContentViewGeneratorCallback;
import com.google.android.libraries.searchinapps.GetSearchContentViewOptions;
import com.google.android.libraries.searchinapps.GetSearchSuggestionsViewGeneratorCallback;
import com.google.android.libraries.searchinapps.GetSearchSuggestionsViewOptions;
import com.google.android.libraries.searchinapps.GetTrendingSearchesViewOptions;
import com.google.android.libraries.searchinapps.LocationContext;
import com.google.android.libraries.searchinapps.LocationContext.CircularArea;
import com.google.android.libraries.searchinapps.LocationContext.GeographicalRestrictions;
import com.google.android.libraries.searchinapps.LocationContext.LatLng;
import com.google.android.libraries.searchinapps.SearchContentViewGenerator;
import com.google.android.libraries.searchinapps.SearchContentViewOptions;
import com.google.android.libraries.searchinapps.SearchInAppsService;
import com.google.android.libraries.searchinapps.SearchSuggestionsViewGenerator;
import com.google.android.libraries.searchinapps.SearchSuggestionsViewOptions;
import com.google.android.material.textfield.TextInputEditText;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/** Android activity demonstrating the usage of SearchInAppsService API. */
public class MainActivity extends AppCompatActivity
    implements GetSearchSuggestionsViewGeneratorCallback {
  static final int MAX_NUM_TRENDS = 3;

  private SearchInAppsService service;
  private SearchInAppsViewModel model;
  private ConstraintLayout suggestionsContainer;
  private LinearLayout searchContentContainer;
  private int searchContentBlock = 0;
  private int searchContentBlocksPerPage = 3;
  private GetSearchContentViewGeneratorCallback searchContentCallback;
  private SearchSuggestionsViewGenerator suggestionsGenerator;
  private SearchContentViewGenerator contentGenerator;

  @Override
  public void onSuccess(SearchSuggestionsViewGenerator generator) {
    model.setSearchSuggestionsViewGenerator(generator);
    suggestionsContainer.removeAllViews();
    suggestionsContainer.addView(generator.populateView(this));
  }

  @Override
  public void onError(String errorMessage) {
    Toast.makeText(this, errorMessage, Toast.LENGTH_SHORT).show();
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    service = SearchInAppsService.create(this);
    model = new ViewModelProvider(this).get(SearchInAppsViewModel.class);
    suggestionsContainer = findViewById(R.id.suggestions_container);
    searchContentContainer = findViewById(R.id.search_content_container);
    RadioGroup radioGroup = findViewById(R.id.layouts_group);
    SearchSuggestionsViewOptions.Layout checkedLayout = model.getCheckedLayout();
    suggestionsGenerator = model.getSearchSuggestionsViewGenerator();
    contentGenerator = model.getSearchContentViewGenerator();

    searchContentCallback =
        new GetSearchContentViewGeneratorCallback() {
          @Override
          public void onSuccess(SearchContentViewGenerator generator) {
            contentGenerator = generator;
            model.setSearchContentViewGenerator(generator);
            searchContentBlock = 0;
            searchContentContainer.removeAllViews();
            while (searchContentBlock < generator.getSearchContentBlockCount()) {
              searchContentContainer.addView(
                  generator.populateView(MainActivity.this, searchContentBlock));
              searchContentBlock++;
            }
          }

          @Override
          public void onLoadMoreSearchContentSuccess(SearchContentViewGenerator generator) {
            while (searchContentBlock < generator.getSearchContentBlockCount()) {
              searchContentContainer.addView(
                  generator.populateView(MainActivity.this, searchContentBlock));
              searchContentBlock++;
            }
          }

          @Override
          public void onLoadMoreSearchContentError(String errorMessage) {
            Toast.makeText(MainActivity.this, errorMessage, Toast.LENGTH_SHORT).show();
          }

          @Override
          public void onError(String errorMessage) {
            Toast.makeText(MainActivity.this, errorMessage, Toast.LENGTH_SHORT).show();
          }
        };
    if (checkedLayout == null) {
      model.setCheckedLayout(
          radioGroup.getCheckedRadioButtonId() == R.id.compact_carousel_layout
              ? COMPACT_CAROUSEL
              : (radioGroup.getCheckedRadioButtonId() == R.id.carousel_layout ? CAROUSEL : TILING));
    }
    findViewById(R.id.suggestion_button)
        .setOnClickListener(
            new View.OnClickListener() {
              @Override
              public void onClick(View v) {
                String textInput =
                    String.valueOf(
                        ((TextInputEditText) findViewById(R.id.suggestion_input)).getText());
                List<String> textContext = Arrays.asList(textInput.split("\\s*,\\s*"));
                GetSearchSuggestionsViewOptions options =
                    new GetSearchSuggestionsViewOptions()
                        .setTextContext(textContext)
                        .setSearchSuggestionsViewOptions(
                            new SearchSuggestionsViewOptions().setLayout(model.getCheckedLayout()));

                String locationInput =
                    String.valueOf(
                        ((TextInputEditText) findViewById(R.id.location_input)).getText());
                if (!locationInput.isEmpty()) {
                  try {
                    options.setLocationContext(
                        Arrays.asList(locationInput.split("\\s*\\|\\s*")).stream()
                            .map(MainActivity::buildLocationContext)
                            .collect(Collectors.toList()));
                  } catch (IllegalArgumentException e) {
                    Toast.makeText(
                            MainActivity.this,
                            "Invalid input for latitude, longitude or radius",
                            Toast.LENGTH_SHORT)
                        .show();
                  }
                }

                service.getSearchSuggestionsView(options, MainActivity.this);
              }
            });
    findViewById(R.id.trending_button)
        .setOnClickListener(
            new View.OnClickListener() {
              @Override
              public void onClick(View v) {
                service.getTrendingSearchesView(
                    new GetTrendingSearchesViewOptions()
                        .setMaxNumTrends(MAX_NUM_TRENDS)
                        .setTrendingSearchesViewOptions(
                            new SearchSuggestionsViewOptions().setLayout(model.getCheckedLayout())),
                    MainActivity.this);
              }
            });
    findViewById(R.id.search_content_button)
        .setOnClickListener(
            new View.OnClickListener() {
              @Override
              public void onClick(View v) {
                String searchRepeatInput =
                    String.valueOf(
                        ((TextInputEditText) findViewById(R.id.search_repeat_input)).getText());
                service.getSearchContentView(
                    new GetSearchContentViewOptions()
                        .setSearchRepeatContext(searchRepeatInput)
                        .setNumberOfBlocksToRequest(searchContentBlocksPerPage)
                        .setSearchContentViewOptions(
                            new SearchContentViewOptions().setLayout(model.getCheckedLayout())),
                    searchContentCallback);
              }
            });
    findViewById(R.id.search_content_more_button)
        .setOnClickListener(
            new View.OnClickListener() {
              @Override
              public void onClick(View v) {
                if (contentGenerator != null) {
                  contentGenerator.loadMoreSearchContent(searchContentCallback);
                }
              }
            });
    if (suggestionsGenerator != null) {
      suggestionsGenerator.getViewOptions().setLayout(model.getCheckedLayout());
      suggestionsContainer.removeAllViews();
      suggestionsContainer.addView(suggestionsGenerator.populateView(this));
    }
    if (contentGenerator != null) {
      contentGenerator.getSearchContentViewOptions().setLayout(model.getCheckedLayout());
      searchContentContainer.removeAllViews();
      while (searchContentBlock < contentGenerator.getSearchContentBlockCount()) {
        searchContentContainer.addView(contentGenerator.populateView(this, searchContentBlock));
        searchContentBlock++;
      }
    }
  }

  public void onRadioButtonClicked(View view) {
    switch (view.getId()) {
      case R.id.carousel_layout:
        model.setCheckedLayout(CAROUSEL);
        break;
      case R.id.tiling_layout:
        model.setCheckedLayout(TILING);
        break;
      default:
        model.setCheckedLayout(COMPACT_CAROUSEL);
    }
    SearchSuggestionsViewGenerator suggestionsGenerator = model.getSearchSuggestionsViewGenerator();
    if (suggestionsGenerator != null) {
      suggestionsGenerator.getViewOptions().setLayout(model.getCheckedLayout());
      suggestionsContainer.removeAllViews();
      suggestionsContainer.addView(suggestionsGenerator.populateView(this));
    }
    SearchContentViewGenerator contentGenerator = model.getSearchContentViewGenerator();
    if (contentGenerator != null) {
      contentGenerator.getSearchContentViewOptions().setLayout(model.getCheckedLayout());
      int childCount = searchContentContainer.getChildCount();
      searchContentContainer.removeViewAt(childCount - 1);
      searchContentContainer.addView(contentGenerator.populateView(this, childCount - 1));
    }
  }

  private static LocationContext buildLocationContext(String locationInput)
      throws IllegalArgumentException {
    List<String> latLngRadius = Arrays.asList(locationInput.split("\\s*,\\s*"));

    if (latLngRadius.size() != 3) {
      throw new IllegalArgumentException();
    }

    return new LocationContext(
        new GeographicalRestrictions(
            new CircularArea(
                new LatLng(
                    Double.parseDouble(latLngRadius.get(0)),
                    Double.parseDouble(latLngRadius.get(1))),
                Integer.parseInt(latLngRadius.get(2)))));
  }

  @Override
  public void onDestroy() {
    service.shutDown();
    super.onDestroy();
  }
}

File: compose/app/src/main/java/com/google/samples/quickstart/searchinapps/compose/MainActivity.kt
/* Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.samples.quickstart.searchinapps.compose

import android.os.Bundle
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.wrapContentWidth
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.google.android.libraries.searchinapps.GetSearchContentViewGeneratorCallback
import com.google.android.libraries.searchinapps.GetSearchContentViewOptions
import com.google.android.libraries.searchinapps.GetSearchSuggestionsViewGeneratorCallback
import com.google.android.libraries.searchinapps.GetSearchSuggestionsViewOptions
import com.google.android.libraries.searchinapps.GetTrendingSearchesViewOptions
import com.google.android.libraries.searchinapps.LocationContext
import com.google.android.libraries.searchinapps.LocationContext.CircularArea
import com.google.android.libraries.searchinapps.LocationContext.GeographicalRestrictions
import com.google.android.libraries.searchinapps.LocationContext.LatLng
import com.google.android.libraries.searchinapps.SearchContentViewGenerator
import com.google.android.libraries.searchinapps.SearchContentViewOptions
import com.google.android.libraries.searchinapps.SearchInAppsService
import com.google.android.libraries.searchinapps.SearchSuggestionsViewGenerator
import com.google.android.libraries.searchinapps.SearchSuggestionsViewOptions

/** Jetpack activity demonstrating the usage of SearchInAppsService API. */
class MainActivity : AppCompatActivity() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContent {
      ApplicationTheme {
        Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
          Column(modifier = Modifier.padding(dimensionResource(R.dimen.column_padding))) {
            SearchSuggestionsUI()
          }
        }
      }
    }
  }

  private fun mapLayout(layout: String): SearchSuggestionsViewOptions.Layout {
    when (layout) {
      "Carousel" -> return SearchSuggestionsViewOptions.Layout.CAROUSEL
      "Tiling" -> return SearchSuggestionsViewOptions.Layout.TILING
      else -> {
        return SearchSuggestionsViewOptions.Layout.COMPACT_CAROUSEL
      }
    }
  }

  @Composable
  fun SearchSuggestionsUI(
    modifier: Modifier = Modifier,
    viewModel: SearchInAppsViewModel = viewModel(),
  ) {
    var context = LocalContext.current
    var service by remember {
      mutableStateOf<SearchInAppsService?>(SearchInAppsService.create(context))
    }
    var textInput by remember { mutableStateOf("") }
    var locationInput by remember { mutableStateOf("") }
    var searchRepeatInput by remember { mutableStateOf("") }

    var searchContentBlockNumber by remember { mutableStateOf(0) }

    val viewSuggestionsGenerator by
      viewModel.getSearchSuggestionsViewGenerator().collectAsStateWithLifecycle()
    val viewContentGenerator by
      viewModel.getSearchContentViewGenerator().collectAsStateWithLifecycle()
    viewContentGenerator?.let { viewContentGenerator ->
      searchContentBlockNumber = viewContentGenerator.getSearchContentBlockCount()
    }
    val layout by viewModel.getLayout().collectAsStateWithLifecycle()

    var radioOptions =
      listOf(
        stringResource(R.string.compact_carousel_layout),
        stringResource(R.string.carousel_layout),
        stringResource(R.string.tiling_layout),
      )

    DisposableEffect(Unit) { onDispose { service?.shutDown() } }

    Column {
      Row(verticalAlignment = Alignment.CenterVertically) {
        OutlinedTextField(
          value = textInput,
          onValueChange = { textInput = it },
          label = { Text(stringResource(R.string.search_hint)) },
          modifier = Modifier.fillMaxWidth(),
        )
      }

      Row(verticalAlignment = Alignment.CenterVertically) {
        OutlinedTextField(
          value = locationInput,
          onValueChange = { locationInput = it },
          label = { Text(stringResource(R.string.location_hint)) },
          modifier = Modifier.fillMaxWidth(),
        )
      }

      Row(verticalAlignment = Alignment.CenterVertically) {
        OutlinedTextField(
          value = searchRepeatInput,
          onValueChange = { searchRepeatInput = it },
          label = { Text(stringResource(R.string.search_repeat_hint)) },
          modifier = Modifier.fillMaxWidth(),
        )
      }

      Spacer(modifier = Modifier.height(dimensionResource(R.dimen.vertical_spacer_height)))

      Row(verticalAlignment = Alignment.CenterVertically) {
        Button(
          modifier = Modifier.wrapContentWidth(),
          onClick = {
            val callback =
              object : GetSearchSuggestionsViewGeneratorCallback {
                override fun onSuccess(generator: SearchSuggestionsViewGenerator) {
                  viewModel.setSearchSuggestionsViewGenerator(generator)
                }

                override fun onError(errorMessage: String) {
                  Toast.makeText(context, errorMessage, Toast.LENGTH_SHORT).show()
                }
              }
            var textContexts: List<String> = listOf<String>(textInput)
            var options: GetSearchSuggestionsViewOptions =
              GetSearchSuggestionsViewOptions()
                .setTextContext(textContexts)
                .setSearchSuggestionsViewOptions(SearchSuggestionsViewOptions().setLayout(layout))

            if (!locationInput.isEmpty()) {
              try {
                options.setLocationContext(
                  locationInput.split("|").map { buildLocationContext(it) }
                )
              } catch (e: IllegalArgumentException) {
                Toast.makeText(
                    this@MainActivity,
                    "Invalid input for latitude, longitude or radius",
                    Toast.LENGTH_SHORT,
                  )
                  .show()
              }
            }

            service?.getSearchSuggestionsView(options, callback)
          },
        ) {
          Text(stringResource(R.string.search_text))
        }

        Spacer(modifier = Modifier.width(dimensionResource(R.dimen.horizontal_spacer_width)))

        Button(
          onClick = {
            val callback =
              object : GetSearchSuggestionsViewGeneratorCallback {
                override fun onSuccess(generator: SearchSuggestionsViewGenerator) {
                  viewModel.setSearchSuggestionsViewGenerator(generator)
                }

                override fun onError(errorMessage: String) {
                  Toast.makeText(context, errorMessage, Toast.LENGTH_SHORT).show()
                }
              }
            var options: GetTrendingSearchesViewOptions =
              GetTrendingSearchesViewOptions()
                .setMaxNumTrends(3)
                .setTrendingSearchesViewOptions(SearchSuggestionsViewOptions().setLayout(layout))
            service?.getTrendingSearchesView(options, callback)
          }
        ) {
          Text(stringResource(R.string.trending_text))
        }
      }

      Spacer(modifier = Modifier.height(dimensionResource(R.dimen.vertical_spacer_height)))

      val searchContentCallback =
        object : GetSearchContentViewGeneratorCallback() {
          override fun onSuccess(generator: SearchContentViewGenerator) {
            viewModel.setSearchContentViewGenerator(generator)
            searchContentBlockNumber = generator.getSearchContentBlockCount()
          }

          override fun onError(errorMessage: String) {
            Toast.makeText(context, errorMessage, Toast.LENGTH_SHORT).show()
          }

          override fun onLoadMoreSearchContentSuccess(generator: SearchContentViewGenerator) {
            searchContentBlockNumber = generator.getSearchContentBlockCount()
          }

          override fun onLoadMoreSearchContentError(errorMessage: String) {
            Toast.makeText(context, errorMessage, Toast.LENGTH_SHORT).show()
          }
        }
      Row(verticalAlignment = Alignment.CenterVertically) {
        Button(
          modifier = Modifier.wrapContentWidth(),
          onClick = {
            var options: GetSearchContentViewOptions =
              GetSearchContentViewOptions()
                .setSearchRepeatContext(searchRepeatInput)
                .setNumberOfBlocksToRequest(3)
                .setSearchContentViewOptions(SearchContentViewOptions().setLayout(layout))

            service?.getSearchContentView(options, searchContentCallback)
          },
        ) {
          Text(stringResource(R.string.search_content))
        }

        Spacer(modifier = Modifier.width(dimensionResource(R.dimen.horizontal_spacer_width)))

        Button(
          onClick = {
            viewContentGenerator?.let { viewContentGenerator ->
              viewContentGenerator.loadMoreSearchContent(searchContentCallback)
            }
          }
        ) {
          Text(stringResource(R.string.load_more_content))
        }
      }

      Spacer(modifier = Modifier.height(dimensionResource(R.dimen.vertical_spacer_height)))

      Row {
        radioOptions.forEach { currentLayout ->
          Row(verticalAlignment = Alignment.CenterVertically) {
            RadioButton(
              selected = (mapLayout(currentLayout) == layout),
              onClick = { viewModel.updateLayout(mapLayout(currentLayout)) },
            )
            Text(text = currentLayout)
          }
        }
      }
      ChipGroupUI(layout, viewSuggestionsGenerator)
      SearchContentUI(layout, viewContentGenerator, searchContentBlockNumber)
    }
  }

  @Composable
  fun ChipGroupUI(
    layout: SearchSuggestionsViewOptions.Layout,
    viewSuggestionsGenerator: SearchSuggestionsViewGenerator?,
  ) {
    viewSuggestionsGenerator?.let { viewSuggestionsGenerator ->
      viewSuggestionsGenerator.getViewOptions().setLayout(layout)
      var context = LocalContext.current
      AndroidView(
        factory = { context -> viewSuggestionsGenerator.populateView(context) },
        update = { view -> viewSuggestionsGenerator.updateView(view, context) },
      )
    }
  }

  @Composable
  fun SearchContentUI(
    layout: SearchSuggestionsViewOptions.Layout,
    viewContentGenerator: SearchContentViewGenerator?,
    searchContentBlockNumber: Int,
  ) {
    viewContentGenerator?.let { viewContentGenerator ->
      viewContentGenerator.getSearchContentViewOptions().setLayout(layout)
      var context = LocalContext.current
      Column(modifier = Modifier.verticalScroll(rememberScrollState())) {
        for (index in 0..searchContentBlockNumber - 1) {
          Row {
            AndroidView(
              factory = { context -> viewContentGenerator.populateView(context, index) },
              update = { view -> viewContentGenerator.updateView(view, context) },
            )
          }
        }
      }
    }
  }

  @Preview(showBackground = true)
  @Composable
  fun SearchSuggestionsPreview() {
    SearchSuggestionsUI()
  }

  private fun buildLocationContext(locationInput: String): LocationContext {
    val latLngRadius = locationInput.split(",")

    require(latLngRadius.size == 3) { "Invalid location input format" }

    return LocationContext(
      GeographicalRestrictions(
        CircularArea(
          LatLng(latLngRadius[0].trim().toDouble(), latLngRadius[1].trim().toDouble()),
          latLngRadius[2].trim().toInt(),
        )
      )
    )
  }
}
